core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java

```diff
@@ -0,0 +1,153 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.CatalogUtil;
+import org.apache.iceberg.hadoop.HadoopConfigurable;
+import org.apache.iceberg.hadoop.SerializableConfiguration;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.SerializableSupplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * FileIO implementation that uses location scheme to choose the correct FileIO implementation.
+ */
+public class ResolvingFileIO implements FileIO, HadoopConfigurable {
+  private static final Logger LOG = LoggerFactory.getLogger(ResolvingFileIO.class);
+  private static final String DEFAULT_SCHEME = "fs";
+  private static final String FALLBACK_IMPL = "org.apache.iceberg.hadoop.HadoopFileIO";
+  private static final Map<String, String> SCHEME_TO_FILE_IO = ImmutableMap.of(
+      DEFAULT_SCHEME, FALLBACK_IMPL,
+      "s3", "org.apache.iceberg.aws.s3.S3FileIO",
+      "s3a", "org.apache.iceberg.aws.s3.S3FileIO",
+      "s3n", "org.apache.iceberg.aws.s3.S3FileIO"
+  );
+
+  private final Map<String, FileIO> ioInstances = Maps.newHashMap();
+  private Map<String, String> properties;
+  private SerializableSupplier<Configuration> hadoopConf;
+
+  @Override
+  public InputFile newInputFile(String location) {
+    return io(location).newInputFile(location);
+  }
+
+  @Override
+  public OutputFile newOutputFile(String location) {
+    return io(location).newOutputFile(location);
+  }
+
+  @Override
+  public void deleteFile(String location) {
+    io(location).deleteFile(location);
+  }
+
+  @Override
+  public void initialize(Map<String, String> newProperties) {
+    close(); // close and discard any existing FileIO instances
+    this.properties = newProperties;
+  }
+
+  @Override
+  public void close() {
+    List<FileIO> instances = Lists.newArrayList();
+
+    synchronized (ioInstances) {
+      instances.addAll(ioInstances.values());
+      ioInstances.clear();
+    }
+
+    for (FileIO io : instances) {
+      io.close();
+    }
+  }
+
+  @Override
+  public void serializeConfWith(Function<Configuration, SerializableSupplier<Configuration>> confSerializer) {
+    this.hadoopConf = confSerializer.apply(hadoopConf.get());
+  }
+
+  @Override
+  public void setConf(Configuration conf) {
+    this.hadoopConf = new SerializableConfiguration(conf)::get;
+  }
+
+  @Override
+  public Configuration getConf() {
+    return hadoopConf.get();
+  }
+
+  private FileIO io(String location) {
+    String impl = implFromLocation(location);
+    FileIO io = ioInstances.get(impl);
```

Comment:
@rdblue Could this be a non-safe operation? The `ioInstances` could be updated during the get. It could be a similar problem as [this](https://stackoverflow.com/questions/17070184/hashmap-stuck-on-get) even if we lock the putting.